lighthouse-core/computed/metrics/cumulative-layout-shift.js

```diff
@@ -0,0 +1,53 @@
+/**
+ * @license Copyright 2019 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const makeComputedArtifact = require('../computed-artifact.js');
+const TraceOfTab = require('../trace-of-tab.js');
+const LHError = require('../../lib/lh-error.js');
+
+class CumulativeLayoutShift {
+  /**
+   * @param {LH.Trace} trace
+   * @param {LH.Audit.Context} context
+   * @return {Promise<LH.Artifacts.MetricValue>}
+   */
+  static async compute_(trace, context) {
+    if (!trace) {
+      throw new Error('Did not provide neccessary data for CLS computation');
+    }
+    const traceOfTab = await TraceOfTab.request(trace, context);
+
+    const layoutShiftEvts = traceOfTab.mainThreadEvents
```

Comment:
nit: if we just need the last one, maybe better just to do
```js
// Find the last LayoutShift event, if any.
let finalLayoutShift;
for (let i = traceOfTab.mainThreadEvents.length; i >= 0; i--) {
  const evt = traceOfTab.mainThreadEvents[i];
  if (evt.name === 'LayoutShift' && evt.args && evt.args.data && evt.args.data.is_main_frame) {
    finalLayoutShift = evt;
    break;
  }
}
```
(they really should make an `Array.findLast()` or whatever)

Reply:
> `evt.args && evt.args.data && evt.args.data.is_main_frame`

optional chaining can't come soon enough :)